src/ray/util/tests/pipe_logger_test.cc

```diff
@@ -137,6 +137,34 @@ TEST(PipeLoggerTestWithTee, RedirectionWithTee) {
   EXPECT_EQ(unlink(test_file_path.data()), 0);
 }
 
+// Testing senario: stream one single content, and have no trailing newliner.
+TEST(PipeLoggerTestWithTee, RedirectionWithTeeAndNoTrailingNewliner) {
+  // TODO(core): We should have a better test util, which allows us to create a temporary
+  // testing directory.
+  const std::string test_file_path = absl::StrFormat("%s.out", GenerateUUIDV4());
+
+  StreamRedirectionOption logging_option{};
+  logging_option.file_path = test_file_path;
+  logging_option.tee_to_stdout = true;
+
+  // Capture stdout via `dup`.
+  testing::internal::CaptureStdout();
+
+  auto stream_redirection_handle = CreateRedirectionFileHandle(logging_option);
+  stream_redirection_handle.CompleteWrite("helloworld", /*len=*/10);
+  stream_redirection_handle.Close();
+
+  // Check content tee-ed to stdout.
+  const std::string stdout_content = testing::internal::GetCapturedStdout();
+  EXPECT_EQ(stdout_content, "helloworld\n");
```

Comment:
It's a bug to have extra newliner, see https://github.com/ray-project/ray/pull/49906#discussion_r1921494519

I plan to address it in a separate PR:
- This one's priority is much higher, since we're losing log data
- I've been working on fixing the newliner for long, still debugging, expect to take more time